Base: fix Unit scheme management regressions

3x380V opened the pull request:
Fixes: #21172
Fixes: #21209
Fixes: #21245
Fixes: #21316
Fixes: #21317

3x380V commented:
@Syres916, thinking more about tests for this issue, we already have `testSchemeTranslation`, but there is no assert for wrong result. It is a shame #21172 sneaked in unnoticed, so we definitely need an assertion here. Added at least `setSchema` / `getSchema` test and more will be added during second round of refactors.

Syres916 commented:
Ok, thanks for the swift fixes.

3x380V commented:
Too bad. I do not have failing environment, so I can only guess. @bofdahof?

Syres916 commented:
@3x380V there appears to still be an issue, if I change the document Units to _Building Euro (cm, m2, m3)_ and run `App.Units.getSchema()` it outputs `2` which would appear to be correct but if I then run `App.Units.schemaTranslate(App.Units.Quantity("1 mm"), App.Units.getSchema())` it gives `('39.37 thou', 0.0254, 'thou')` which is definitely wrong.

A similar incorrect result when _Imperial for Civil Eng (ft, ft/s)_ is selected `App.Units.getSchema()` it outputs `6` but `App.Units.schemaTranslate(App.Units.Quantity("1 mm"), App.Units.getSchema())` gives `('1.00 mm', 1.0, 'mm')` again incorrect.

Syres916 commented:
OK, it didn't catch my attention last night but the order of the document Units has changed so the statement I made about _Building Euro (cm, m2, m3)_

> `App.Units.getSchema()` it outputs `2` which would appear to be correct 

is in fact not correct, it should output `4`. Here is the list as previous set:

![Old_Units_Order](https://github.com/user-attachments/assets/d0aab0a7-2fe3-43de-8b9c-94b1f3586bad)

and here is the new:

![New_Units_Order](https://github.com/user-attachments/assets/895f50ee-ab51-4680-8d0c-f87344e428a5)

but importantly the _references_ haven't changed, see https://github.com/FreeCAD/FreeCAD/blob/main/src/Base/UnitsSchemasData.h#L540 (the `4` at the start of the line is the correct reference for _Building Euro (cm, m2, m3)_ so the getSchema() is still the root cause, if I use `App.Units.schemaTranslate(App.Units.Quantity("1 mm"), 4)` I get the correct output of `('0.10 cm', 10.0, 'cm')`

3x380V commented:
@Syres916, let me start with a disclaimer, I'm not the author of that change, just a messenger. Now, the problem appears to be a confusion between internal schema number and the array index. So that `2` is the schema number and means Imperial schema, so conversion is correct. The obvious fix is to drop conversion to the schema numbers and use array index directly. @bofdahof coded a whole machinery around this, probably to maintain compatibility, so I'd try to wait a while for his input.

3x380V commented:
Ah, I guess, I see the intention. Once sorted against internal indexes, everything should work.

3x380V commented:
@bofdahof, could you look at this issue? I do understand magic behind `UnitsSchemas::getVec` and `schemaSpecs {s3, s4, s5, s6, s7, s8, s9, s0, s1, s2}`, but I do not understand the intention. So I might as well make the things worse.

@Syres916, while there, do you know why is the second argument of `schemaTranslate` mandatory? The obvious thing here would be to do conversion using currently selected schema.

Syres916 commented:
> @Syres916, while there, do you know why is the second argument of `schemaTranslate` mandatory? The obvious thing here would be to do conversion using currently selected schema.

No sorry, I don't know the history of this part of the code.

3x380V commented:
@bofdahof, if `getVec` returns vector sorted according `UnitsSchemaSpec::num`, then the Python interface would work again. I'm just clueless, why do we need two indexes, `num` and `schemaSpecs`'s

3x380V commented:
Hmm, even C++ code is wrong the very same way Python interface is:
https://github.com/FreeCAD/FreeCAD/blob/9938818d82df5efcb347b29615f52d42355b2cc2/src/Gui/PreferencePages/DlgSettingsGeneral.cpp#L121-L125

So we have index into description vector and later is used as

https://github.com/FreeCAD/FreeCAD/blob/9938818d82df5efcb347b29615f52d42355b2cc2/src/Gui/PreferencePages/DlgSettingsGeneral.cpp#L224-L229

but `setSchema` does expect schema number, which is a very different thing.

3x380V commented:
@bofdahof, all the users, as it is implemented now, relies on the order of `schemaSpecs`. The code selecting schema would magically start working once sorted according `num`. That's why I was repeatedly asking what is an idea behind it. As it stands now, implementation is broken both for Python and C++. Question is whenever you care about fixing that or I'm on my own ;-)

3x380V commented:
@bofdahof, thanks! That's exactly the information I needed. The opposite direction would be to drop `num` and rely on the order given in the vector (that's what I would do without your input). Your preferred solution would need some check there are no holes in the indexes and no duplicities to be long term reliable. But it is the easiest fix for this bug, so I will implement it.

3x380V commented:
I won't as explained previously. But let me take an opportunity for a short remark: this situation is a shining example why small, isolated, incremental changes are a good thing. Without your input, dropping `num` and using index into `schemaSpecs` vector would do the equivalent job. After all, you need to define how schemas follow one after another. That's both for Python's `tuple` interface (which you cannot change as an external (workbench) code may rely on it) and for all the combo boxes, etc. Or perhaps you do not need to do that on `UnitsApi` level, but then those interface methods, `names()` and `descriptions()` should return both `string` and `num` so the caller knows how to select a particular scheme using its index (num). And as noone can read your mind, smaller isolated changes would help to sort things out in case something went wrong.

3x380V commented:
Just as a side note. CAM workbench is using creative constructs as
```
unitstring = "imperial" if FreeCAD.Units.getSchema() in [2, 3, 5, 7] else "metric"
```
or
```
minute_based_schemes = list(map(FreeCAD.Units.listSchemas, [6, 3, 2]))
```

3x380V commented:
@Syres916, please test.

@sliptonic, some CAM tests depend on default unit schema being selected, but they do not set it explicitly. Perhaps something to improve.

Syres916 commented:
@3x380V so far nothing's producing any errors, things like Attachment Editor open swiftly and as expected and the correct schema is being picked up in the various parts of the code. Thanks for efforts, it's appreciated.

The only thing from a user's perspective is if I had `Standard` (num=0) from version 1.0.x or below and open a file with a TechDrawing instead of _100_ (mm) shown as a dimension I'll now get _0.1_ (m) under `Internal` so I'd have to change all my docs to `Metric small parts` (num=6). Not a massive job and I don't think it should stop this PR from being merged but not sure @WandererFan is aware??

Syres916 commented:
@3x380V do you want to cherry pick the changes from https://github.com/FreeCAD/FreeCAD/compare/main...Syres916:FreeCAD:Revert_Internal_metres_to_Standard_millimetres into this PR?

3x380V commented:
@Syres916, lets fix that in single PR, but as I'm traveling it is a bit painful do to serious testing. Therefore I dropped test from my commit and added your fix only. This way this PR could be merged quickly while another one could be open for tests.

I'm assuming those two commits for `GeneralSettingsWidget` in your branch should be squashed. I guess the most straightforward way is that you open PR with tests once this one gets merged.

Syres916 commented:
> I'm assuming those two commits for `GeneralSettingsWidget` in your branch should be squashed. I guess the most straightforward way is that you open PR with tests once this one gets merged.

Yep, that's fine, I was using the web interface but I'll squash them remaining into a new branch.

3x380V commented:
So are we done here and this PR can be merged? It is a pity weekly build was released without these fixes...

3x380V commented:
@bofdahof, I took pragmatic approach here. That does not mean I'm happy with that, but it solves regressions reported. I suggest revisiting this after #21147 gets merged (see especially unit definition order problem described in https://github.com/FreeCAD/FreeCAD/pull/21147#discussion_r2084139656).

@hyarion (or anyone else with merge rights), gentle ping. I will rebase #21147 on top of this, to ease testing.

hyarion commented:
@3x380V @Syres916 @bofdahof thanks for looking into this
